17 fix the readme badges and update the reade me fle (#19)
FAIRmat-NFDI/pynxtools-stm: README.md

[![Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/astral-sh/ruff/main/assets/badge/v2.json)](https://github.com/astral-sh/ruff)
![](https://github.com/FAIRmat-NFDI/pynxtools-xrd/actions/workflows/pytest.yml/badge.svg)
![](https://github.com/FAIRmat-NFDI/pynxtools-xrd/actions/workflows/pylint.yml/badge.svg)
![](https://github.com/FAIRmat-NFDI/pynxtools-xrd/actions/workflows/publish.yml/badge.svg)
![](https://img.shields.io/pypi/pyversions/pynxtools-xrd)
![](https://img.shields.io/pypi/l/pynxtools-xrd)
![](https://img.shields.io/pypi/v/pynxtools-xrd)
![](https://coveralls.io/repos/github/FAIRmat-NFDI/pynxtools_xrd/badge.svg?branch=master)

# STS reader

> ***Note: Though the reader name is STS reader, it also supports STM experiment species. This is the first version of the reader according to the NeXus application definition [NXsts](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml) which is a generic template of concepts' definition for STS and STM experiments. Later on, the application definitions and readers specific to the STM, STS and AFM will be available. To stay upto date keep visiting this page time to time. From now onwards we will mention STS referring both STM and STS.***

Reviewer: Please specify here what these acronyms mean.

Author: Done!

> ***Note: Though the reader name is STS reader, it also supports STM experiment species. This is the first version of the reader according to the NeXus application definition [NXsts](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml) which is a generic template of concepts' definition for STS and STM experiments. Later on, the application definitions and readers specific to the STM, STS and AFM will be available. To stay upto date keep visiting this page time to time. From now onwards we will mention STS referring both STM and STS.***

Reviewer: ```suggestion
***Note: Though the reader name is STS reader, it also supports STM experiment species. This is the first version of the reader according to the NeXus application definition [NXsts](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml) which is a generic template of concept definitions for STS and STM experiments. Later on, the application definitions and readers specific to STM, STS and AFM will be available. To stay up-to-date, keep visiting this page from time to time. From now onwards, when we mention STS, we are referring to both STM and STS.***
```

Author: Done!

> ***Note: Though the reader name is STS reader, it also supports STM experiment species. This is the first version of the reader according to the NeXus application definition [NXsts](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml) which is a generic template of concept definitions for STS and STM experiments. Later on, the application definitions and readers specific to STM, STS and AFM will be available. To stay up-to-date, keep visiting this page from time to time. From now onwards, when we mention STS, we are referring to both STM and STS.***

Author: Noted!

> The prime purpose of the reader is to transform lab-defined data into community-defined concepts constructed by the SPM community which allows experimentalists to store, organize, search, analyze, and share experiments data (only within of [NOMAD](https://nomad-lab.eu/nomad-lab/) platform) among the scientific communities. To utilize the reader one needs 

Reviewer: ```suggestion
The prime purpose of the reader is to transform data from measurement files into community-defined concepts constructed by the SPM community which allows experimentalists to store, organize, search, analyze, and share experiments data (only within the [NOMAD](https://nomad-lab.eu/nomad-lab/) platform) among the scientific communities. To utilize the reader one needs 
```

Author: Done!

> The prime purpose of the reader is to transform data from measurement files into community-defined concepts constructed by the SPM community which allows experimentalists to store, organize, search, analyze, and share experiments data (only within the [NOMAD](https://nomad-lab.eu/nomad-lab/) platform) among the scientific communities. To utilize the reader one needs 

Author: Noted!

1. A data file from the experiment

> 2. Config file (to map the NeXus concepts and the raw data coming from the experimental data file)

Reviewer: ```suggestion
2. A config file (to map the raw data coming from the experimental data file to the corresponding NeXus concepts
```

Author: Done!

> 3. An ELN (Electronic Lab Notebook) file (to add user-provided data that does not come along with the experimental data file)

Author: Noted!

- Can parse Scanning Tunneling Microscopy (STM) from
    - Nanonis: Generic 5e, Generic 4.5

> ## Introduction to the Different Input Files

Reviewer: I would put the experimental data file first

### ELN (Electronic Lab Notebook)

> This file mainly supports user input data. There are two ways to define or write ELN files, The First one can be distinguished, for sake of explanation, as command line ELN, a yaml file with `.yaml`. Such type of ELN needs to be used to run the reader from command line. The second one can be called, for sake of explanation, Nomad Schema ELN is a yaml file with extension `.scheme.archive.yaml`. The ELN is needed if the reader is being used from Nomad. Note that Nomad will parser Nomad Schema ELN into a yaml file of the first type.

Reviewer: ```suggestion
This file supports user input data that is not part of the experimental data. There are two ways to define or write ELN files. The first one can be distinguished, for sake of explanation, as command line ELN, a yaml file with `.yaml`. Such type of ELN needs to be used to run the reader from the command line. The second one can be called, for sake of explanation, Nomad Schema ELN is a yaml file with extension `.scheme.archive.yaml`. This ELN is needed if the reader is being used from Nomad. Note that Nomad will parse Nomad Schema ELN into a yaml file of the first type.
```

Author: Done!

            type_data:
              - sts
              - stm
          m_annotations:
            eln:
              component: EnumEditQuantity
          description: 'Only two type of experiments are allowed: sts and stm.'
      sub_sections:
        Instrument:
          section:
            m_annotations:
              eln:
                overview: true
            quantities:
              stm_head_temp:
                type: np.float64
                unit: K
                m_annotations:
                  eln:
                    component: NumberEditQuantity
                    defaultDisplayUnit: K
                description: |
                  Temperature of STM head. Note: At least one field from stm_head_temperature,
                  cryo_bottom_temp and cryo_sheild_temp must be provided. '
        sample:
          section:
            m_annotations:
              eln:
                overview: true
            quantities:
              name:
                type: str
                m_annotations:
                  eln:
                    component: StringEditQuantity
                description: |
                  Name of the sample.
```

> The `section`, `sub_sections`, and `quantities` refer to root level entitiy (behaibes like a `group`), `group`, and `field` of the NeXus definition. The given schema eln can be read as follows, `stm` ELN has direct fields `default`, `definition` and direct groups `Insturment`, `Sample`. Where each group can have nested `group`s and `field`s.

Reviewer: ```suggestion
The `section`, `sub_sections`, and `quantities` refer to root level entitiy (behaves like a `group`), `group`, and `field` of the NeXus definition. The given schema eln can be read as follows, `stm` ELN has direct fields `default`, `definition` and direct groups `Instrument`, `Sample`. Where each group can have nested `group`s and `field`s.
```

This sentence needs to be cleaned up

Author: Done!

The example given below is to explain the **command line ELN**.
```yaml
Instrument:
  hardware:
    name: Nanonis
  lock_in:
    lock_in_data_flip_number: -1.0
    modulation_amplitude:
      unit: V
      value: 0.005
    modulation_frequency:
      unit: Hz
      value: 973.0
    status: 'OFF'
  piezo_config:
    active_calib: LHe
  sample_bias:
    bias:
      unit: V
      value: 0.005
    bias_calibration: 1.0
    bias_offset:
      unit: V
      value: '0'
  software:
    rt_release: '10771'
    ui_release: '10771'
    vendor: nanonis
    version: Generic 5e
  stm_head_temp:
    unit: K
    value: 5.04866
collection_identifier: TiSe2_2303a_annealing_300C_5min_evaporate_Pyrene_1_
default: backward
definition: NXsts
entry_identifier: TiSe2_2303a_annealing_300C_5min_evaporate_Pyrene_1_0070
experiment_description: A new TiSe2, annealed at 300 C for 5 min, then cool down to
  RT, evaporate the Pyrene on RT, 2.2 E -7, totally 10 s.
experiment_identifier: C:\Users\SPM-PEEM\Desktop\DATA_Nanonis\20220711_CreaTec_Service_Benchmarks_LHe\Nanonis-Session-PMD100-HVHU_CreaTec_Service_PalmaLabBerlin220711
experiment_type: stm
type: background
```

> This type of ELN one needs to fill to use as a input file for the reader. To know what fields and groups refering what type data one needs to read the NeXus definition in [FAIRmat NeXus Proposal](https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXsts.html#nxsts) page or in [github repo](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml). 

Reviewer: ```suggestion
This type of ELN needs to be used if the reader is run from the comman line. To know what fields and groups refer which type of data, one needs to read the NeXus definition on the [FAIRmat NeXus Proposal](https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXsts.html#nxsts) page or in the [GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml). 
```

Author: Done!

### Config File

> Config file is used to map NeXus definition concepts to the raw data coming from sts experiment.

Reviewer: ```suggestion
The config file is used to map the raw data coming from STS experiment to the NeXus definition concepts.
```

                                   "/dat_mat_components/Temperature 1 [filt]/metadata",
                                   "/dat_mat_components/LI Demod 1 X [filt]/value",
                                   "/dat_mat_components/LI Demod 1 X [filt]/unit",
                                   "/dat_mat_components/LI Demod 1 X [filt]/metadata",
                                   "/dat_mat_components/LI Demod 1 Y [filt]/value",
                                   "/dat_mat_components/LI Demod 1 Y [filt]/unit",
                                   "/dat_mat_components/LI Demod 1 Y [filt]/metadata",
                                   "/dat_mat_components/LI Demod 2 X [filt]/value",
                                   "/dat_mat_components/LI Demod 2 X [filt]/unit",
                                   "/dat_mat_components/LI Demod 2 X [filt]/metadata",
                                   "/dat_mat_components/LI Demod 2 Y [filt]/value",
                                   "/dat_mat_components/LI Demod 2 Y [filt]/unit",
                                   "/dat_mat_components/LI Demod 2 Y [filt]/metadata"]}
}
```
**NOTES**

> - To update (if needed) the config file please follow the rules:

Reviewer: ```suggestion
- To update (if needed) the config file, a set of rules needs to be followed:
```

>   - The dictionary in config files have the following meaning?

Reviewer: ```suggestion
  - The dictionary in config files have the following meaning:
```

- Can parse Scanning Tunneling Microscopy (STM) from
    - Nanonis: Generic 5e, Generic 4.5

## Introduction to the Different Input Files
### Raw File

> This type of file (such as `example.dat`, `example.sxm`) just come from the experiment. 

Author: Noted!

### ELN (Electronic Lab Notebook)

> This file supports user input data that is not part of the experimental data file. There are two ways to define or write ELN files. The first one can be distinguished, for sake of explanation, as command line ELN, a yaml file with `.yaml`. Such type of ELN needs to be used to run the reader from the command line. The second one can be called, for sake of explanation, Nomad Schema ELN is a yaml file with extension `.scheme.archive.yaml`. This ELN is needed if the reader is being used from Nomad. Note that Nomad will parse Nomad Schema ELN into a yaml file of the first type.

Author: Noted!

> The given example below is a short description of the **Nomad schema ELN** (for getting a full example please visit [here](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-remote-tools-hub/-/blob/develop/docker/sts/example/sts_example/STS.scheme.archive.yaml?ref_type=heads)).

Author: Noted!

            type_data:
              - sts
              - stm
          m_annotations:
            eln:
              component: EnumEditQuantity
          description: 'Only two type of experiments are allowed: sts and stm.'
      sub_sections:
        Instrument:
          section:
            m_annotations:
              eln:
                overview: true
            quantities:
              stm_head_temp:
                type: np.float64
                unit: K
                m_annotations:
                  eln:
                    component: NumberEditQuantity
                    defaultDisplayUnit: K
                description: |
                  Temperature of STM head. Note: At least one field from stm_head_temperature,
                  cryo_bottom_temp and cryo_sheild_temp must be provided. '
        sample:
          section:
            m_annotations:
              eln:
                overview: true
            quantities:
              name:
                type: str
                m_annotations:
                  eln:
                    component: StringEditQuantity
                description: |
                  Name of the sample.
```

> The `section`, `sub_sections`, and `quantities` refer to root level entitiy (behaves like a `group`), `group`, and `field` of the NeXus definition. The given schema eln can be read as follows, `stm` ELN has direct fields `default`, `definition` and direct groups `Instrument`, `Sample`. Where each group can have nested `group`s and `field`s.

Author: Noted!

    bias_offset:
      unit: V
      value: '0'
  software:
    rt_release: '10771'
    ui_release: '10771'
    vendor: nanonis
    version: Generic 5e
  stm_head_temp:
    unit: K
    value: 5.04866
collection_identifier: TiSe2_2303a_annealing_300C_5min_evaporate_Pyrene_1_
default: backward
definition: NXsts
entry_identifier: TiSe2_2303a_annealing_300C_5min_evaporate_Pyrene_1_0070
experiment_description: A new TiSe2, annealed at 300 C for 5 min, then cool down to
  RT, evaporate the Pyrene on RT, 2.2 E -7, totally 10 s.
experiment_identifier: C:\Users\SPM-PEEM\Desktop\DATA_Nanonis\20220711_CreaTec_Service_Benchmarks_LHe\Nanonis-Session-PMD100-HVHU_CreaTec_Service_PalmaLabBerlin220711
experiment_type: stm
type: background
```
This type of ELN needs to be used if the reader is run from the comman line. To know what fields and groups refer which type of data, one needs to read the NeXus definition on the [FAIRmat NeXus Proposal](https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXsts.html#nxsts) page or in the [GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml). 
### Config File

> Config file is used to map NeXus definition concepts to the raw data coming from sts experiment and user input data.

Author: Noted!

**NOTES**

- To update (if needed) the config file, a set of rules needs to be followed:
  - The dictionary in config files have the following meaning:
    ```
    "/ENTRY[entry]/INSTRUMENT[instrument]/lock_in/harmonic_order_N": {"D1": {"value": "/Lock-in/Harmonic D1/value"},
                                                                      "D2": {"value": "/Lock-in/Harmonic D2/value"}},
    ```
    Here, the part `N` in field `harmonic_order_N`, can be considered as the name of dimensions, can be replaced by `D1` and `D2` to  write two `harmonic_order` and can be extended for further dimensions.
  - List for the same concept
    ```
    "/ENTRY[entry]/INSTRUMENT[instrument]/piezo_config/active_calib": ["/Piezo Configuration/Active Calib.",
                                                                       "/Piezo Calibration/Active Calib."],
    ```
    For different type of software versions the raw data path could be different for the same concept. For example, Nanonis software `generic 5e` has `/Piezo Configuration/Active Calib.` and generic 4.5 has `/Piezo Calibration/Active Calib.` for the same concept `/ENTRY[entry]/INSTRUMENT[instrument]/piezo_config/active_calib`.

>   - In the config file the concept that takes data from ELN is denoted by `@eln` otherwise data will come from experimental raw files.

Author: Nonted!

>   - Importantly one should note the `NXdata` concept `/ENTRY[entry]/DATA[data]` it takes dict of the list. Each key (`0`, `1` ...) of the dict refers to a NXdata group of data fields `bias` and `current` for multiple given setup, saying, with and without `filter` check points. For other setup, one can extend the dict following the same convenstion used here.

Author: Noted!

## Run the Reader from CLI

> In command line cli input files come as positional arguments and the other comes a keyword argument.

Reviewer: ```suggestion
In the command line interface, input files can be passed as positional arguments. Other arguments like the `reader` and the `nxdl` shall be given as keyword arguments.
```

Author: Noted!

### Run STS Reader
The following command can be used to run the `STS` reader:
```bash
!dataconverter \
--reader sts \
--nxdl NXsts \
$<path-to STS_nanonis_generic_5e_1.dat> \
$<path-to config_file_for_dat.json> \
$<path-to Nanonis_Eln.yaml> \
--output ./final_sts_dev.nxs
```

### Run STM Reader

> Use the following command to run the `STM` reader

Reviewer: ```suggestion
Use the following command to run the `STM` reader:
```

Author: Noted!

```bash
!dataconverter \
--reader sts \
--nxdl NXsts \
$<path-to STM_nanonis_generic_5e.sxm> \
$<path-to config_file_for_sxm.json> \
$<path-to Nanonis_Eln.yaml> \
--output final_stm_dev_.nxs
```

> ## Want to Contribute or Extend the Reader Functionality

Reviewer: You have this point now twice, once as `## Want to Contribute or Extend the Reader Functionality` and once as `### How to contribute / How to extend the reader functionality` (line 284 ff.). Please consider merging.

Author: Noted!

We appreciate any comments, extentions or improvements on the existing reader from users. Currently the reader supports the versions `4.5`, `5e` from `Generic` model of `Nanonis` vendor. To include the other versions of the `Generic` model, extend the class `StmNanonisGeneric` and `StsNanonisGeneric` by including versions in `__version__` attribute. Also include the model and version of the brand in `Spm` class. 
### How to contribute / How to extend the reader functionality
If you want to add the different versions of the `Nanonis Generic` model for `STM` experiment, please check the `STM_Nanonis` class for `STM` experiment, which parses the data into a python dict of slash (`/`) separated key-value pair (see the right part of the config file). The class uses the `nanonispy` sub-package to read the `sxm` type file. That should also be checked and modified (if needed).

If you add different versions of the `Nanonis Generic` model for `STS` experiment, please check the `BiasSpecData_Nanonis` class for `STS` experiment. The class reads the raw data from the raw files into a dict of slash separated key-value pair (see the config file). 

> If you go for a completely defferent model from the different brands, please handle it in a new module with different functions and classes. 

Author: Noted!

> Later on please add the relavent tests in the plugin test.

Reviewer: ```suggestion
Later on, please add the relevant tests in the plugin test.
```

Author: Noted!

Done! Great, then please create a pull request.

## Useful Functions:

> There are a few usages of functions that you can utilze to fit this reader compatible to you data. 

Reviewer: ```suggestion
There are a few functions that you can utilize to make this reader compatible with your data. 
```

Author: NOted!

    For different type of software versions the raw data path could be different for the same
    concept. For example, Nanonis software `generic 5e` has `/Piezo Configuration/Active Calib.`
    and generic 4.5 has `/Piezo Calibration/Active Calib.` for the same concept `/ENTRY[entry]/INSTRUMENT[instrument]/piezo_config/active_calib`.

>   - In the config file the concept that takes data from ELN is denoted by `@eln` otherwise data will come from experimental raw files.

Reviewer: ```suggestion
  - In the config file the concept that takes data from ELN is denoted by the token `@eln`. Otherwise, data will come from raw experimental files.
```

Author: Done!

>   - Importantly one should note the `NXdata` concept `/ENTRY[entry]/DATA[data]` it takes dict of the list. Each key of the dict refers to the `bais` and `current` in multiple given setup, saying, with and without `filter` check points. For other setup, one can extend the dict following the same convenstion used here.

Reviewer: ```suggestion
  - Importantly, the `NXdata` concept `/ENTRY[entry]/DATA[data]` takes a list of dict. Each key of the dict refers to the `bias` and `current` in multiple given setup, saying, with and without `filter` check points. For another setup, one can extend the dict following the same convenstion used here.
```
Not sure what you wanted to say here.

Author: Done!

### Raw File
This type of file (such as `example.dat`, `example.sxm`) just come from the experiment. 

> ## Run the Reader from CLI

Reviewer: Move this up so people know how to use this tool.

Author: Done!

In command line cli input files come as positional arguments 
### Run STS Reader
The following command can be used to run the sts reader:
```bash
!dataconverter \
--reader sts \
--nxdl NXsts \
$<path-to STS_nanonis_generic_5e_1.dat> \
$<path-to config_file_for_dat.json> \
$<path-to Nanonis_Eln.yaml> \
--output ./final_sts_dev.nxs
```

### Run STM Reader
Use the following command to run the stm reader
```bash
!dataconverter \
--reader sts \
--nxdl NXsts \
$<path-to STM_nanonis_generic_5e.sxm> \
$<path-to config_file_for_sxm.json> \
$<path-to Nanonis_Eln.yaml> \
--output final_stm_dev_.nxs
```

> ## Want to Contribiute or Extend the Reader Functionality

Reviewer: ```suggestion
## How to contribute / How to extend the reader functionality
```

> We appreciate any comments, extentions or improvements on the existing reader from users. Current the reader supports the versions `4.5`, `5e` from `Generic` model of `Nanonis` vendor. To include the other versions of the `Generic` model, extend the class `StmNanonisGeneric` and `StsNanonisGeneric` by including versions in `__version__` attribute. At the same time also include the model and version of the brand in `Spm` class. 

Reviewer: ```suggestion
We appreciate any comments, extentions, or improvements on the existing reader from users. Currently, the reader supports the versions `4.5`, `5e` from `Generic` model of `Nanonis` vendor. To include any other versions of the `Generic` model, extend the class `StmNanonisGeneric` and `StsNanonisGeneric` by including versions in `__version__` attribute. At the same time, you need to also include the model and version of the brand in `Spm` class. 
```

> If you add the different versions of the `Nanonis Generic` model for `STM` experiment, please check the `STM_Nanonis` class for `STM` experiment, which parse the data into python dict of slash (`/`) separated-key and value pair (see the right part of the confic file). The class usages the `nanonispy` sub-package to read the `sxm` type file. That should also be check and modified.

Reviewer: ```suggestion
If you want to add different versions of the `Nanonis Generic` model for `STM` experiment, please check the `STM_Nanonis` class for `STM` experiment, which parses the data into a python dict of slash (`/`) separated key-value pairs (see the right part of the config file). The class uses the `nanonispy` sub-package to read the `sxm` type file. That should also be checked and (if needed) modified.
```

Author: Done!

> If you add the different versions of the `Nanonis Generic` model for `STS` experiment, please check the `BiasSpecData_Nanonis` class for `STS` experiment. The class reads the raw data from the raw files into a slash separated-key and value pair (see the config file). 

Reviewer: ```suggestion
If you add a different versions of the `Nanonis Generic` model for `STS` experiment, please check the `BiasSpecData_Nanonis` class for `STS` experiment. The class reads the raw data from the raw files into a dict of slash separated key-value pairs (see the config file). 
```

Author: Done!

If you go for a completely defferent model from the different brands, please handle it in a new module with different functions and classes. 

Later on please add the relavent tests in the plugin test.

Done! Great, then please create a pull request.

## Useful Functions:
There are a few usages of functions that you can utilse to fit this reader compatible to you data. 

> - **Function get_stm_raw_file_info()**: For `STM` experiment the function can return you the slash separated dict in a text fle. This dict helps to write or modify the config file according to you raw data file. 

Reviewer: ```suggestion
- **Function get_stm_raw_file_info()**: For `STM` experiment the function can return you the slash separated dict in a text file. This dict helps to write or modify the config file according to your raw data file. 
```

Author: Done!
